File: server/server-core/src/main/java/dzida/server/core/basic/unit/Point.java
package dzida.server.core.basic.unit;

import java.util.Objects;

public final class Point {
    private final double x;
    private final double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public static Point of(double x, double y) {
        return new Point(x, y);
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public Point plus(Point position) {
        return new Point(this.x + position.x, this.y + position.y);
    }

    public Point minus(Point position) {
        return new Point(this.x - position.x, this.y - position.y);
    }

    public Point multiply(double scalar) {
        return new Point(this.x * scalar, this.y * scalar);
    }

    public Point divide(double scalar) {
        return new Point(this.x / scalar, this.y / scalar);
    }

    public double distanceTo(Point position) {
        return Math.sqrt(this.distanceSqrTo(position));
    }

    public double distanceSqrTo(Point position) {
        double dx = position.x - this.x;
        double dy = position.y - this.y;
        return dx * dx + dy * dy;
    }

    public boolean isInRange(Point position, double radius) {
        return distanceSqrTo(position) <= radius * radius;
    }

    public double crossProduct(Point p) {
        return x * p.y - y * p.y;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Point position = (Point) o;
        return Double.compare(position.x, x) == 0 &&
                Double.compare(position.y, y) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ')';
    }
}

File: client/public/src/unit/point.js
define(function (require, exports, module) {

    function Point(x, y) {
        this.x = x;
        this.y = y;
    }

    function interpolate(ratio, v1, v2) {
        const dv = v2 - v1;
        return v1 + dv * ratio;
    }

    Point.fromObject = function (data) {
        return new Point(data.x, data.y);
    };

    Point.distanceSqr = (p1, p2) => {
        const dx = p2.x - p1.x;
        const dy = p2.y - p1.y;
        return dx * dx + dy * dy;
    };

    Point.distance = (p1, p2) => {
        return Math.sqrt(Point.distanceSqr(p1, p2));
    };

    Point.isInRange = (p1, p2, radius) => Point.distanceSqr(p1, p2) <= radius * radius;

    Point.interpolate = function (ratio, p1, p2) {
        var x = interpolate(ratio, p1.x, p2.x);
        var y = interpolate(ratio, p1.y, p2.y);
        return {x, y}
    };

    Point.angleFromTo = (p1, p2) => {
        const dx = p2.x - p1.x;
        const dy = p2.y - p1.y;
        return -Math.atan2(dx, dy);
    };

    Point.multiplyInPlace = (p, scalar) => {
        p.x *= scalar;
        p.y *= scalar;
        return p;
    };

    Point.equal = (p1, p2) => {
        return p1.x == p2.x && p1.y == p2.y;
    };

    Point.prototype.toString = function () {
        return '(' + this.x + ', ' + this.y + ')';
    };

    Point.prototype.plus = function (point) {
        return new Point(this.x + point.x, this.y + point.y);
    };

    Point.prototype.minus = function (point) {
        return new Point(this.x - point.x, this.y - point.y);
    };

    Point.prototype.multiply = function (scalar) {
        return new Point(this.x * scalar, this.y * scalar);
    };

    Point.prototype.divide = function (scalar) {
        return this.multiply(1.0 / scalar);
    };

    Point.prototype.dotProduct = function (point) {
        return this.x * point.x + this.y * point.y;
    };

    Point.prototype.dotProductSqr = function () {
        return this.x * this.x + this.y * this.y;
    };

    Point.prototype.crossProduct = function (point) {
        return this.x * point.y - this.y * point.y;
    };

    Point.prototype.distanceTo = function (point) {
        return Math.sqrt(this.distanceSqrTo(point));
    };

    Point.prototype.distanceSqrTo = function (point) {
        return Point.distanceSqr(this, point)
    };

    Point.prototype.isInRange = function (point, radius) {
        return Point.isInRange(this, point, radius)
    };

    module.exports = Point;
});